### PR TITLE
sensors::VtkUtil: Add VtkUtil class and the unit test

### PR DESCRIPTION
**Pull request by kunimatsu-tri**
Created the utility class for VTK.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5035)

### CODE REVIEW

**Comment by liangfok** [comment]
+@liangfok 

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/vtk_util.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0qfM87-JVo2IdgtWk:-Kc0qfM87-JVo2IdgtWl:b-1jeqj0) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L13)):*
> ```objc
> class VtkUtil {
>  public:
>   /// Create a square plane with the given size, color and:
> ```

Per style guide, "Create" should be "Creates".

---

*[drake/systems/sensors/vtk_util.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0qvyIiJE5Z5QE90Ra:-Kc0qvyIiJE5Z5QE90Rb:b-dcoht5) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L14)):*
BTW, Consider rewording

> the z axis to be the plane normal

to be:

> the plane's z axis is its normal and is coincident with the z axis of the world coordinate system.

---

*[drake/systems/sensors/vtk_util.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0xMFsoKko4Hp67qQJ:-Kc0xMFsoKko4Hp67qQK:bfu0kao) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L14)):*
> ```objc
>  public:
>   /// Create a square plane with the given size, color and:
>   ///   - the z axis to be the plane normal
> ```

BTW, what about the plane's x and y axes?

---

*[drake/systems/sensors/vtk_util.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0r4RNCoO7tx13HJC-:-Kc0r4ROZsgHlCGV0ei3:bzej0wj) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L15)):*
BTW, Consider rewording

> - the plane center at (0, 0, 0).

to be:

> - the plane's center is at (0, 0, 0) of the world coordinate system.

---

*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc0tcya8g8b1Ok9PEm9:bzggx6q) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
BTW, the use of `const unsigned char` is likely a style guide violation:

https://google.github.io/styleguide/cppguide.html#Integer_Types

> You should not use the unsigned integer types such as uint32_t, unless there is a valid reason such as representing a bit pattern rather than a number, or you need defined overflow modulo 2^N. In particular, do not use unsigned types to say a number will never be negative. Instead, use assertions for this.

Can you pass in a `const int color[3]` instead?

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0uspXHqKluniRd_Oa:-Kc0uspXHqKluniRd_Ob:b657fqv) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L54)):*
> ```cpp
> 
>  private:
>   std::vector<Point> point_set_;
> ```

BTW, should `#include <vector>` per style guide.

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0ui_RjKpKebJ-WS5W:-Kc0ui_RjKpKebJ-WS5X:b-gm7a29) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L58)):*
> ```cpp
> 
> 
> TEST_F(VtkUtilTest, PointsCorrespoindenceTest) {
> ```

BTW, consider adding a brief comment describing what this unit test is doing.

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0vC4lw_OaZxPKAHMB:-Kc0vC4lw_OaZxPKAHMC:b-terkcu) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L59)):*
> ```cpp
> 
> TEST_F(VtkUtilTest, PointsCorrespoindenceTest) {
>   unsigned char color[3] = {255, 128, 0};
> ```

Can this be declared const?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-:-Kc0wvbGgx1l3afg4rvp:b-gyxe3j)*

**Comment by kunimatsu-tri** [comment]
Review status: 3 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/sensors/vtk_util.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0qfM87-JVo2IdgtWk:-Kc4M-flNoR6OeRDNHnJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L13)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Per style guide, "Create" should be "Creates".
</blockquote></details>

Done.

---

*[drake/systems/sensors/vtk_util.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0xMFsoKko4Hp67qQJ:-Kc4J3ueoZn0rnR5GTyJ:b-e7n23w) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L14)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, what about the plane's x and y axes?
</blockquote></details>

They are also the same as world coordinate system's.  Changed the sentence to describe this.

---

*[drake/systems/sensors/vtk_util.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0qvyIiJE5Z5QE90Ra:-Kc4M-x5C53zMx2AJJdv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L14)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, Consider rewording

> the z axis to be the plane normal

to be:

> the plane's z axis is its normal and is coincident with the z axis of the world coordinate system.
</blockquote></details>

Done.

---

*[drake/systems/sensors/vtk_util.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0r4RNCoO7tx13HJC-:-Kc4M0ws0MwXEDNHHjJ7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, Consider rewording

> - the plane center at (0, 0, 0).

to be:

> - the plane's center is at (0, 0, 0) of the world coordinate system.
</blockquote></details>

Done.

---

*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc4KLn04QofCkzSTMPF:bah76yc) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, the use of `const unsigned char` is likely a style guide violation:

https://google.github.io/styleguide/cppguide.html#Integer_Types

> You should not use the unsigned integer types such as uint32_t, unless there is a valid reason such as representing a bit pattern rather than a number, or you need defined overflow modulo 2^N. In particular, do not use unsigned types to say a number will never be negative. Instead, use assertions for this.

Can you pass in a `const int color[3]` instead?
</blockquote></details>

Actually, this is what VTK needs.  Perhaps, need to document?

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0uspXHqKluniRd_Oa:-Kc4M1sD7JCXD29iRyoO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L54)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should `#include <vector>` per style guide.
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0ui_RjKpKebJ-WS5W:-Kc4M26gHnan8qQhxuC7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L58)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a brief comment describing what this unit test is doing.
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/vtk_util_test.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0vC4lw_OaZxPKAHMB:-Kc4M2neuvdaMjWHAJvM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/test/vtk_util_test.cc#L59)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Can this be declared const?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc4N6da9-HxfgWe5GB3:b57kctz) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Actually, this is what VTK needs.  Perhaps, need to document?
</blockquote></details>

BTW, can this method take a `const int color[3]` but then internally convert it into a `const unsigned char color[3]` possibly with some corner-case error checking like verifying all of the integer values are between 0 and 255? That would prevent leakage of the non-style-guide-conforming type beyond the scope of this method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/vtk_util.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc4NgOiX5T2lMr35ZR9:-Kc4NgOjcYiX0cigPZe8:b-tr9tbx) ([raw file](https://github.com/robotlocomotion/drake/blob/f812b5ed8bc690de38704e131fbbc353afff13eb/drake/systems/sensors/vtk_util.cc#L23)):*
> ```cpp
> 
>   vtkNew<vtkPolygon> polygon;
>   polygon->GetPointIds()->SetNumberOfIds(4);  // Make a quad
> ```

BTW, missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by kunimatsu-tri** [comment]
*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc4P-vY-03gKWalZ-XW:b-b0smnt) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, can this method take a `const int color[3]` but then internally convert it into a `const unsigned char color[3]` possibly with some corner-case error checking like verifying all of the integer values are between 0 and 255? That would prevent leakage of the non-style-guide-conforming type beyond the scope of this method.
</blockquote></details>

SGTM. will do so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by kunimatsu-tri** [comment]
*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc4PPbUfffvbJNymIP3:b-6te8dx) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

SGTM. will do so.
</blockquote></details>

Ah, no, it doesn't work for this case.
I need to directly pass the array of unsigned char to VTK's api.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
*[drake/systems/sensors/vtk_util.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0tcya8g8b1Ok9PEm8:-Kc4QrgpZcprU1r7d3oW:b3ec7zs) ([raw file](https://github.com/robotlocomotion/drake/blob/8ce4827e8af89e523b927c4f87e364e4c44a4f2c/drake/systems/sensors/vtk_util.h#L21)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Ah, no, it doesn't work for this case.
I need to directly pass the array of unsigned char to VTK's api.
</blockquote></details>

That's too bad. Oh well, let's see what the other reviewers say. I'll mark myself satisfied for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by kunimatsu-tri** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/vtk_util.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc4NgOiX5T2lMr35ZR9:-Kc4Ru-7jmqAYoZtGces:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f812b5ed8bc690de38704e131fbbc353afff13eb/drake/systems/sensors/vtk_util.cc#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing period at end of sentence.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/sensors/vtk_util.h, line 10 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc4UEZ-QZKtGp00RXz_:-Kc4UEZ0PMtqW90fiTnf:b1fzqpj) ([raw file](https://github.com/robotlocomotion/drake/blob/aa1bf8ac9d2003da02fd846b5c3e99e33227cf7d/drake/systems/sensors/vtk_util.h#L10)):*
> ```objc
> namespace sensors {
> 
> /// Utility class for VTK library.
> ```

BTW: nit: missing "the" after "for".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

File: drake/systems/sensors/CMakeLists.txt [hunk 1]
```diff
@@ -24,7 +24,7 @@ target_link_libraries(drakeSensors
     drakeSystemFramework)
 
 if(VTK_FOUND)
-  target_link_libraries(drakeSensors vtkRendering)
+  target_link_libraries(drakeSensors vtkRendering vtkIO)
```
Comment (patmarion) [inline drake/systems/sensors/CMakeLists.txt]: are you actually using vtkIO in this PR?

**Comment by patmarion** [comment]
LGTM.  By the way, VTK provides a vtkPlaneSource class which can make the plane geometry for you.  If you decided to use that class to generate the geometry, then your vtk util class could include a general method to add vertex colors to the points of any vtkPolyData object.

**Comment by kunimatsu-tri** [comment]
@patmarion Thanks.  I tried vtkPlaneSource, but I couldn't find the way to change the plane size with it. It always returns 1.0m x 1.0m plane. That's why I implemented by my own.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-:-Kc4egaIcDbr1IAbzyE4:b-6pl34x)*

**Comment by kunimatsu-tri** [comment]
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/CMakeLists.txt, line 27 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc0JDFuM9K1iFDpWJSc-r3-27:-Kc4gJYRToGe9J7_to6p:b-gbi77g) ([raw file](https://github.com/robotlocomotion/drake/blob/aa1bf8ac9d2003da02fd846b5c3e99e33227cf7d/drake/systems/sensors/CMakeLists.txt#L27)):*
<details><summary><i>Previously, patmarion (Pat Marion) wrote…</i></summary><blockquote>

are you actually using vtkIO in this PR?
</blockquote></details>

Oops. This was for the next PR.

---

*[drake/systems/sensors/vtk_util.h, line 10 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc4UEZ-QZKtGp00RXz_:-Kc4gfmEc2Sog3Hqlbn-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa1bf8ac9d2003da02fd846b5c3e99e33227cf7d/drake/systems/sensors/vtk_util.h#L10)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: missing "the" after "for".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by liangfok** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-:-Kc4k0KUW6S6zsGf1vhd:bnfp4nl)*

**Comment by kunimatsu-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-:-Kc4mBHH13Vw6Z1R9kW2:bg0xcbm)*

**Comment by patmarion** [comment]
@kunimatsu-tri btw, you can experiment with vtk in the Python console of drake-visualizer (press F8 or click the button in the toolbar to open the console).  For example, here is how to use vtkPlaneSource to visualize a plane with vtkPlaneSource:

```
halfWidth = 0.5
p = vtk.vtkPlaneSource()
p.SetOrigin(-halfWidth, -halfWidth, 0)
p.SetPoint1(halfWidth, -halfWidth, 0)
p.SetPoint2(-halfWidth, halfWidth, 0)
p.Update()
vis.showPolyData(p.GetOutput(), 'plane')
```

**Comment by kunimatsu-tri** [comment]
Cool! What is vis here? I couldn't execute "import vis".

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-:-Kc4oK4Yg_nyTsgeuCNS:b-vf1adr)*

**Comment by patmarion** [comment]
vis should already be imported in drake-visualizer (at least in recent editions), but if you need to import it then: `from director import visualization as vis`

**Comment by liangfok** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5035#-Kc4yWCaWHvt168HhfSX:-Kc4yWCbPlVSfltemWGv:b3vcfut) (no related file):*
BTW, this PR has some `cpplint` issues:

```
cd drake-distro
./drake/common/test/cpplint_wrapper.py
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5035)*

**Comment by kunimatsu-tri** [comment]
vtkPlaneSource should be used instead of this.

**Pull request closed without merging**